File: erlang/word-count/word_count.erl
-module(word_count).
-export([count/1]).

count([], Map) ->
  dict:from_list(maps:to_list(Map));
count([W|T], Map) ->
  count(T, maps:put(W, maps:get(W, Map, 0) + 1, Map)).
count(String) ->
  Cleaned = re:replace(String,"[^A-Za-z0-9]+", " ", [{return, list}]),
  Stripped = string:strip(Cleaned),
  Normed = string:to_lower(Stripped),
  Tokens = re:split(Normed, "[^a-z0-9]+", [trim,{return, list}]),
  count(Tokens, #{}).

File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map()
  def count(sentence) do
    map = %{}
    Enum.each(tokenize(sentence), fn word -> value = map[word]; Dict.put(map, word, value || 0 + 1) end)

    map
  end

  defp tokenize(sentence) do
    cleaned = Regex.replace(~r/[^A-Za-z0-9]+/, sentence, " ")
    stripped = String.strip(cleaned)
    normed = String.downcase(stripped)
    Regex.split(~r/[^a-z0-9]+/, normed)
  end
end
